### ISSUE TITLE
Thread safety issues (was: calling glslopt_initialize/glslopt_cleanup many times)

### ISSUE THREAD

**Issue by eodabash** [aras-p/glsl-optimizer#76]
We build the glsl optimizer source directly into an ios app and (perhaps not the best practice) create a new glslopt_ctx for each shader we compile at runtime. Recently (possibly as of Xcode 6) we seem to be dying sometimes inside glslopt_cleanup with memory errors, or glslopt_optimize starts failing with a lot of error messages about seemingly correct code. 

Before I do a lot of digging to try to narrow down a test case I wondering if creating multiple glslopt contexts per process/GL context/whatever is even safe to do, or if there is global state being stored somewhere.

**Comment by volcoma** [comment]
I can confirm this is true. I have the same problem when used with multiple threads. Is it using some kind of shared state or buffer inside?

**Comment by eodabash** [comment]
I think the issue is this stuff in builtin_functions.cpp

`//@TODO: implement
typedef int mtx_t;
#define _MTX_INITIALIZER_NP 0
#define mtx_lock(name)
#define mtx_unlock(name)`

Plus these globals in glsl_types.cpp

`hash_table *glsl_type::array_types = NULL;
hash_table *glsl_type::record_types = NULL;
hash_table *glsl_type::interface_types = NULL;
void *glsl_type::mem_ctx = NULL;`

If I supply an implementation for the first group and add similar synchronization around the second, then test case I posted originally doesn't crash anymore. I believe this is all mesa code so it might be worth checking if they've made any improvements to thread safety in the compiler since glsl optimizer was last updated.

**Comment by aras-p** [comment]
Indeed, it very well might be that there's some part that is not thread safe somewhere. Did not have time to dig in & find it though.
